### ISSUE TITLE
[ch03] 요약 및 정리

### ISSUE THREAD

**Issue by JaehoNam** [JaehoNam/aas-practice#4]
## 3장. 음악 추천과 Audioscrobbler 데이터셋

So, it seems fitting to kick off the next three chapters, which will explore key machine learning algorithms on Spark, with a chapter built around recommender engines, and recommending music in particular. It’s an accessible way to introduce real-world use of Spark and MLlib, and some basic machine learning ideas that will be developed in subsequent chapters.(음악에 특화된 추천 엔진 공부)

### 3.1 데이터셋

This example will use a data set published by Audioscrobbler.(오디오스크로블러 제공)
The Audioscrobbler data set is interesting because it merely records plays.(단순 재생 정보만 기록 -> 좋아하는지 싫어하는지 판단 불가능)
A data set likethis is therefore much larger, covers more users and artists, and contains more totalinformation than a rating data set, even if each individual data point carries less information. This type of data is often called implicit feedback data because the user-artist connections are implied as a side effect of other actions, and not given as explicit ratings or thumbs-up.(암묵적 피드백, 정보의 총량이 많음)
The main dataset is in the user_artist_data.txt file. It contains about 141,000 unique users, and 1.6million unique artists. About 24.2 million users’ plays of artists are recorded, alongwith their counts.

### 3.2 교차 최소 제곱 추천 알고리즘

We need to choose a recommender algorithm that is suitable for this implicit feed‐back data. The data set consists entirely of interactions between users and artists’ songs. It contains no information about the users, or about the artists other than their names. We need an algorithm that learns without access to user or artist attributes. These are typically called collaborative filtering algorithms.(협업 필터링 알고리즘 -> 두 사람이 들은 노래 중 같은 것이 많기 때문에 같은 노래를 좋아할 수도 있다고 예측)

Finally, we need an algorithm that scales, both in its ability to build large models and to create recommendations quickly. Recommendations are typically required in nearreal time—within a second, not tomorrow.(대규모 & 빠른 추천 알고리즘)

This example will employ a member of a broad class of algorithms called latent-factor models. They try to explain observed interactions between large numbers of users and items through a relatively small number of unobserved, underlying reasons.(다수의 아이템 -> 적은 수의 숨은 원인으로 설명)

Fortunately, there are algorithms that can escape this catch-22 and find a decent solution. More specifically still, the example in this chapter will use the Alternating Least Squares (ALS) algorithm to compute X and Y.(교차 최소 제곱 알고리즘)

Finally, the ALS algorithm can take advantage of the sparsity of the input data as well. This, and its reliance on simple, optimized linear algebra and its data-parallel nature, make it very fast at large scale. This is much of the reason it is the topic of this chapter—that, and the fact that ALS is the only recommender algorithm currently imple‐mented in Spark MLlib!(ALS 알고리즘의 장점)

### 3.3 데이터 준비하기

However, the map() function must return exactly one value for every input, so it can’tbe used. It’s possible to remove the lines that don’t parse with filter(), but this would duplicate the parsing logic. The flatMap() function is appropriate when each element maps to zero, one, or more results because it simply “flattens” these collections of zero or more results from each input into one big data set. It works with Scala Collections, but also with Scala’s Option class. Option represents a value that might only optionally exist. It is like a simple collection of 1 or 0 values, corresponding to its Some and None subclasses.(map 은 에러가 있는 줄을 넘길 수 없고, filter 는 전체를 2번 훑어야 함 -> flatmap 에 some/none 을 넣으면 알아서 0 또는 1 길이의 iterable 을 flat 하는 구조)

### 3.4 첫 번째 모델 만들기

Instead, we create a broadcast variable called bArtistAlias for artistAlias. This makes Spark send and hold in memory just one copy for each executor in the cluster. When there are thousands of tasks and many execute in parallel on each executor, this can save significant network traffic and memory.(여러 곳에서 동시 사용되는 map variable 을 broadcast 에 등록)

Broadcast variables are useful when many tasks need access to the same (immutable) data structure. They extend normal handling of task closures to enable:
- Caching data as raw Java objects on each executor, so they need not be deserialized for each task
- Caching data across multiple jobs, stages, and tasks

SparkML 의 ALS model 을 가져와서 학습시킴, seed 를 랜덤으로 제공해주어야 매번 다른 결과를 얻을 수 있음
`setDefault(seed, this.getClass.getName.hashCode.toLong)` -> 고정 시드

### 3.5 추천 결과 추출 검사하기

```
  def makeRecommendations(model: ALSModel, userID: Int, howMany: Int): DataFrame = {
    val toRecommend = model.itemFactors.
      select($"id".as("artist")).
      withColumn("user", lit(userID))
    model.transform(toRecommend).
      select("artist", "prediction").
      orderBy($"prediction".desc).
      limit(howMany)
  }
```
위의 함수에서 transform 부분이 직관적으로 이해가 되지 않아서 소스 코드를 분석해보았다.

```
  @Since("2.0.0")
  override def transform(dataset: Dataset[_]): DataFrame = {
    transformSchema(dataset.schema)
    // create a new column named map(predictionCol) by running the predict UDF.
    val predictions = dataset
      .join(userFactors,
        checkedCast(dataset($(userCol))) === userFactors("id"), "left")
      .join(itemFactors,
        checkedCast(dataset($(itemCol))) === itemFactors("id"), "left")
      .select(dataset("*"),
        predict(userFactors("features"), itemFactors("features")).as($(predictionCol)))
    getColdStartStrategy match {
      case ALSModel.Drop =>
        predictions.na.drop("all", Seq($(predictionCol)))
      case ALSModel.NaN =>
        predictions
    }
  }
```
먼저 transformSchema 를 통해 prediction column 을 추가한다. 다음으로 val predictions 를 만드는 과정은 먼저 join 을 통해 user id, item id 에 해당하는 userFactors.features, itemFactors.features column 을 추가한 뒤 select 에서 `dataset("*")` 을 통해 userCol, itemCol 에 해당하는 column 을 select 하고 userFactors.features, itemFactors.features column 을 통해 predictionCol column 을 새로 추가한다. predict 는 udf(user defined function) 으로 정의되어 있다. userFactors("features") 는 userFactors.features 와 같은 의미이다.

### 3.6 추천 품질 평가하기

This metric is directly related to an information retrieval concept called the receiver operating characteristic (ROC) curve. The metric in the preceding paragraph equals the area under this ROC curve, and is indeed known as AUC, or Area Under the Curve. AUC may be viewed as the probability that a randomly chosen good recommendation ranks above a randomly chosen bad recommendation.(ROC curve, AUC 정의, ROC 는 x 축이 FP rate, y 축이 TP rate 를 의미하기 때문에 curve 가 작은 x 쪽으로, 큰 y 쪽으로 몰려 있을 때 모델의 정확도가 높다고 할 수 있다. 따라서 AUC 가 1에 가까울수록 좋다.)

Other evaluation metrics that are relevant to systems that rank things are implemented in RankingMetrics. These include metrics like precision, recall, and mean average precision (MAP). MAP is also frequently used and focuses more narrowly on the quality of the top recommendations. However, AUC will be used here as a common and broad measure of the quality of the entire model output.(다른 metric 들도 존재)

### 3.7 AUC 계산하기

In order to use it, we must split the input data into a training and CV set. The ALS model will be trained on the training data set only, and the CV set will be used to evaluate the model. Here, 90% of the data is used for training and the remaining 10% for cross-validation.(train, cv set 으로 나누기, test set 은 나중에)

The result is about 0.879. Is this good? It is certainly higher than the 0.5 that is expected from making recommendations randomly, and it’s close to 1.0, which is the maximum possible score. Generally, an AUC over 0.9 would be considered high.(내가 해 봤을 때는 AUC 값이 0.9 였고 model 은 만들 때마다 seed 에 의해 달라지기 때문이다.)

### 3.8 하이퍼파라미터 선택하기

The configured hyperparameters were:
- setRank(10)
  - The number of latent factors in the model, or equivalently, the number of columns k in the user-feature and product-feature matrices. In nontrivial cases, this is also their rank.(latent factor 개수)
- setMaxIter(5)
  - The number of iterations that the factorization runs. More iterations take more time but may produce a better factorization.(반복 횟수)
- setRegParam(0.01)
  - A standard overfitting parameter, also usually called lambda. Higher values resist overfitting, but values that are too high hurt the factorization’s accuracy.(overfitting 조절 변수, 값이 클수록 overfitting 방지)
- setAlpha(1.0)
  - Controls the relative weight of observed versus unobserved user-product interactions in the factorization.(이미 관측된 값에 집중하는게 좋으면 alpha 값을 크게 하면 좋음, ALS 논문에서는 40을 기본값으로 잡아둠)

### 3.9 추천 결과 만들기

Interestingly, this entire process could also be used to recommend users to artists. This could be used to answer questions like “Which 100 users are most likely to be interested in the new album by artist X?” Doing so would only require swapping the user and artist field when parsing the input.(user, artist 순서만 바꾸면 artist 에게 user 도 추천 가능)

### 3.10 한 걸음 더 나아가기
